numjuggler/nogq2.py

```diff
@@ -1,444 +1,451 @@
-    else:
-        ares = B - A <= atol
-        astr = f2.format(atol)
-    if rtol is None:
-        rres = False
-        rstr = f1.format(rtol)
-    else:
-        b = max(map(abs, l))
-        rres = B - A <= rtol*b
-        rstr = f2.format(rtol*b)
-
-    result = ares or rres
-
-    if cmnt is not None:
-        # assume it is a list of comments. Add here information
-        c = cmnt.append
-        c('Are close check: ' + name + ': {}'.format(result))
-        if detailed:
-            c('  values: ' + ' '.join('{:15.8e}'.format(v) for v in l))
-            c('  B:      {:15.8e}'.format(B))
-            c('  A:      {:15.8e}'.format(A))
-            c('  B - A:  {:15.8e}'.format(B - A))
-            c('  atol:   ' + astr)
-            c('  rtol*b: ' + rstr)
-    return result
-
-
-def get_params(card):
-    """
-    Return parameters of the GQ card.
-    """
-    p1, p2 = card.lower().split('gq')
-    # Check if there is transformation:
-    tr = len(p1.split()) > 1
-    return tr, map(float, p2.split())
-
-
-def get_cone_or_cyl(pl):
-    """
-    See notes from 14.03.2018
-    """
-    cmnt = []
-
-    # Define normalization coeff gamma:
-    A, B, C, D, E, F, G, H, J, K = pl
-    ABC = sum((A, B, C))
-    cmnt.append(' A, B, C:' + ' '.join('{:15.8e}'.format(v) for v in (A, B, C)))
-    cmnt.append(' D, E, F:' + ' '.join('{:15.8e}'.format(v) for v in (D, E, F)))
-    cmnt.append(' G, H, J:' + ' '.join('{:15.8e}'.format(v) for v in (G, H, J)))
-    cmnt.append('       K:' + ' {:15.8e}'.format(K))
-
-    # List of normalization coefficients. 1 is always assumed
-    gammas = [1.0, -1.0]
-    tDEF = 5e-5
-    tABC = 1e-9
-    if areclose((0, E), atol=tDEF, rtol=None):
-        if areclose((1, B), atol=tABC, rtol=None):
-            g = 1.0/B
-            gammas.append(g)
-            cmnt.append('gamma_B = {:15.8e}'.format(g))
-        if areclose((1, C), atol=tABC, rtol=None):
-            g = 1.0/C
-            gammas.append(g)
-            cmnt.append('gamma_C = {:15.8e}'.format(g))
-    else:
-        g = 2*E/(2*E*A - D*F)
-        gammas.append(g)
-        cmnt.append('gamma_E = {:15.8e}'.format(g))
-
-    if areclose((0, F), atol=tDEF, rtol=None):
-        if areclose((1, A), atol=tABC, rtol=None):
-            g = 1.0/A
-            gammas.append(g)
-            cmnt.append('gamma_A = {:15.8e}'.format(g))
-        if areclose((1, C), atol=tABC, rtol=None):
-            g = 1.0/C
-            gammas.append(g)
-            cmnt.append('gamma_C = {:15.8e}'.format(g))
-    else:
-        g = 2*F/(2*F*B - D*E)
-        gammas.append(g)
-        cmnt.append('gamma_F = {:15.8e}'.format(g))
-
-    if areclose((0, D), atol=tDEF, rtol=None):
-        if areclose((1, B), atol=tABC, rtol=None):
-            g = 1.0/B
-            gammas.append(g)
-            cmnt.append('gamma_B = {:15.8e}'.format(g))
-        if areclose((1, A), atol=tABC, rtol=None):
-            g = 1.0/A
+from __future__ import print_function, division, nested_scopes
```

Comment:
``git diff `` does not show properly the introduced changes. I suppose, the original code is wrapped with the ``if six.PY2 ...`` clause. This will result in preventing the ``--mode nogq2`` to work under Python3. Why?

Reply:
The module relies on vector class from pirs. The pirs library is not PY3 compatible and I had some problems with installing it even in PY2 environment. I couldn't devote more time for this, so this is just quick fix. If you don't have pirs, you don't have nogq2.

By the way, why don't you use numpy as basis for this math? Numpy is nearly ubiquitous in python environments, so, it's not a heavy dependency.

Reply:
Ok, agree.

> ... why don't you use numpy ...

Historically. For 5 years numpy was not so simple to install on a linux cluster into the user environment (now anaconda makes this simple).